File: backend/src/controllers/ParishController.php
<?php

namespace App\Controllers;

use App\Controllers\Traits\JsonResponseTrait;
use App\Models\Parish;
use Psr\Http\Message\ResponseInterface as Response;
use Psr\Http\Message\ServerRequestInterface as Request;

class ParishController
{
    use JsonResponseTrait;

    /**
     * GET /parishes
     */
    public function getAll(Request $request, Response $response): Response
    {
        try {
            $parishes = Parish::with('diocese')->get();
            return $this->jsonResponse($response, 200, true, "All parishes fetched", $parishes);
        } catch (\Exception $e) {
            return $this->jsonResponse($response, 500, false, "Error fetching parishes", [
                "details" => $e->getMessage()
            ]);
        }
    }

    /**
     * GET /parishes/{id}
     */
    public function getById(Request $request, Response $response, array $args): Response
    {
        try {
            $parish = Parish::with('diocese')->find($args['id']);
            if (!$parish) {
                return $this->jsonResponse($response, 404, false, "Parish not found");
            }
            return $this->jsonResponse($response, 200, true, "Parish fetched successfully", $parish);
        } catch (\Exception $e) {
            return $this->jsonResponse($response, 500, false, "Error fetching parish", [
                "details" => $e->getMessage()
            ]);
        }
    }

    /**
     * POST /parishes
     */
    public function create(Request $request, Response $response): Response
    {
        try {
            $data = json_decode($request->getBody(), true);

            // Minimal validation
            if (empty($data['DioceseID']) || empty($data['ParishName']) || empty($data['StateID'])) {
                return $this->jsonResponse($response, 400, false, "Missing one or more required fields: DioceseID, ParishName, StateID");
            }

            $parish = Parish::create($data);
            return $this->jsonResponse($response, 201, true, "Parish created successfully", $parish);
        } catch (\Exception $e) {
            return $this->jsonResponse($response, 500, false, "Error creating parish", [
                "details" => $e->getMessage()
            ]);
        }
    }

    /**
     * PUT /parishes/{id}
     */
    public function update(Request $request, Response $response, array $args): Response
    {
        try {
            $parish = Parish::find($args['id']);
            if (!$parish) {
                return $this->jsonResponse($response, 404, false, "Parish not found");
            }

            $data = json_decode($request->getBody(), true);
            $parish->update($data);

            return $this->jsonResponse($response, 200, true, "Parish updated successfully", $parish);
        } catch (\Exception $e) {
            return $this->jsonResponse($response, 500, false, "Error updating parish", [
                "details" => $e->getMessage()
            ]);
        }
    }

    /**
     * DELETE /parishes/{id}
     */
    public function delete(Request $request, Response $response, array $args): Response
    {
        try {
            $parish = Parish::find($args['id']);
            if (!$parish) {
                return $this->jsonResponse($response, 404, false, "Parish not found");
            }

            $parish->delete();
            return $this->jsonResponse($response, 204, true, "Parish deleted successfully");
        } catch (\Exception $e) {
            return $this->jsonResponse($response, 500, false, "Error deleting parish", [
                "details" => $e->getMessage()
            ]);
        }
    }
}

File: backend/src/main/java/com/nearchrist/backend/controller/ParishController.java
package com.nearchrist.backend.controller;

import com.nearchrist.backend.dto.ParishDto;
import com.nearchrist.backend.dto.ParishUpsertDto;
import com.nearchrist.backend.dto.ApiResponse;
import com.nearchrist.backend.service.ParishService;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/parishes")
public class ParishController {
    private final ParishService service;

    public ParishController(ParishService service) {
        this.service = service;
    }

    @GetMapping
    public ResponseEntity<ApiResponse<List<ParishDto>>> getAll() {
        try {
            List<ParishDto> parishes = service.getAllParishes();
            return ResponseEntity.ok(new ApiResponse<>(true, 200, "All parishes fetched", parishes));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(new ApiResponse<>(false, 500, "Error fetching parishes: " + e.getMessage(), null));
        }
    }

    @GetMapping("/{id}")
    public ResponseEntity<ApiResponse<ParishDto>> getById(@PathVariable Long id) {
        try {
            return service.getParishById(id)
                    .map(p -> ResponseEntity.ok(new ApiResponse<>(true, 200, "Parish fetched successfully", p)))
                    .orElseGet(() -> ResponseEntity.status(HttpStatus.NOT_FOUND)
                            .body(new ApiResponse<>(false, 404, "Parish not found", null)));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(new ApiResponse<>(false, 500, "Error fetching parish: " + e.getMessage(), null));
        }
    }

    @PostMapping
    public ResponseEntity<ApiResponse<ParishDto>> create(@RequestBody ParishUpsertDto dto) {
        try {
            ParishDto saved = service.createParish(dto);
            return ResponseEntity.status(HttpStatus.CREATED)
                    .body(new ApiResponse<>(true, 201, "Parish created successfully", saved));
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest()
                    .body(new ApiResponse<>(false, 400, e.getMessage(), null));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(new ApiResponse<>(false, 500, "Error creating parish: " + e.getMessage(), null));
        }
    }

    @PutMapping("/{id}")
    public ResponseEntity<ApiResponse<ParishDto>> update(@PathVariable Long id, @RequestBody ParishUpsertDto dto) {
        try {
            return service.updateParish(id, dto)
                    .map(p -> ResponseEntity.ok(new ApiResponse<>(true, 200, "Parish updated successfully", p)))
                    .orElseGet(() -> ResponseEntity.status(HttpStatus.NOT_FOUND)
                            .body(new ApiResponse<>(false, 404, "Parish not found", null)));
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest()
                    .body(new ApiResponse<>(false, 400, e.getMessage(), null));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(new ApiResponse<>(false, 500, "Error updating parish: " + e.getMessage(), null));
        }
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<ApiResponse<Void>> delete(@PathVariable Long id) {
        try {
            if (!service.deleteParish(id)) {
                return ResponseEntity.status(HttpStatus.NOT_FOUND)
                        .body(new ApiResponse<>(false, 404, "Parish not found", null));
            }
            return ResponseEntity.ok(new ApiResponse<>(true, 204, "Parish deleted successfully", null));
        } catch (IllegalStateException e) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                    .body(new ApiResponse<>(false, 400, e.getMessage(), null));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(new ApiResponse<>(false, 500, "Error deleting parish: " + e.getMessage(), null));
        }
    }
}